### ISSUE TITLE
Feature] Framework-level fine-grained `torch.compile` + Mega Cache for Omni models

### ISSUE THREAD

**Issue by zhaochenyang20** [sgl-project/sglang-omni#172]
> Supersedes #172. Based on benchmarking discussion in [PR #153](https://github.com/sgl-project/sglang-omni/pull/153) (thanks [@Ratish1](https://github.com/Ratish1) for the detailed profiling).

## Motivation

`torch.compile` delivers meaningful throughput improvements for omni models — up to **+37% tok/s** on S2-Pro's codebook decoder with partial compile alone. However, it currently isn't used because:

1. **Startup overhead**: Full-model compile takes 2–5 minutes depending on model size, unacceptable for production cold starts.
2. **No framework-level abstraction**: Any compile logic today would have to be hard-coded into each model's factory, duplicating boilerplate and mixing compile concerns with model logic.
3. **CUDA graph interaction**: The framework already manages CUDA graph capture. `torch.compile` must coexist cleanly, not fight for control.

This issue proposes a **three-phase plan** that introduces `torch.compile` incrementally at the framework level, so **every omni model** (S2-Pro, Qwen3-Omni, future models) can opt in via configuration without touching model code.

## Architecture context

The `sglang-omni` engine stack:

```
OmniEngine
  ├─ Scheduler (model-agnostic: request lifecycle, batching)
  │     ├─ BatchPlanner        (pluggable, per-model)
  │     ├─ ResourceManager     (pluggable, per-model)
  │     └─ IterationController (pluggable, per-model)
  └─ ModelRunner (stateless executor)
        ├─ model(**inputs)     ← torch.compile target
        ├─ InputPreparer       (pluggable, per-model)
        └─ OutputProcessor     (pluggable, per-model)
```

Models expose compile-friendly sub-modules as **compile targets** — self-contained functions/modules that models register, and the framework wraps with `torch.compile` at engine creation time.

**Current models and their compile-relevant structure**:

| Model | Backbone (slow head) | Auxiliary module (fast head) | Compile opportunity |
|---|---|---|---|
| **S2-Pro** | Qwen3 transformer (24L) via RadixAttention | Codebook decoder (`_decode_codebooks`: project_in + 10× forward_kvcached + argmax) | Aux module: +37% tok/s. Full model: +43% |
| **Qwen3-Omni** | Qwen3 thinker via SGLang AR | Talker (speech token generation), audio/image encoders | Encoders and talker are self-contained compile targets |
| **Future models** | Any SGLang-backed LLM | Model-specific decoders, post-processors | Should get compile support for free |

## Benchmark baseline (S2-Pro, from PR #153)

[@Ratish1](https://github.com/Ratish1) benchmarked three configurations:

| Configuration | Startup time | Steady-state throughput | Δ throughput |
|---|---|---|---|
| CUDA graph only (current) | ~33s | 88 tok/s | baseline |
| Partial compile (fast head) | ~54s | 121 tok/s | **+37%** |
| Full-model compile | ~137s | 126 tok/s | +43% |

Key takeaway: **partial compile captures ~85% of the total compile gain at ~20% of the startup cost.** The transformer backbone is already well-served by CUDA graph; `torch.compile` adds kernel fusion wins primarily in auxiliary modules with small matmuls, sequential loops, and embedding lookups.

## Design principles

1. **No `torch.compile()` calls inside model files.** Models declare what is compilable; the framework decides whether and how to compile.
2. **Compilable targets must be self-contained.** Tensor-in, tensor-out. No `self` attribute access in the compiled path — pass everything as arguments.
3. **`fullgraph=True` is mandatory.** Graph breaks defeat the purpose. If a function can't compile as a full graph, fix the function, don't relax the constraint.
4. **Eager-first readability.** The code path must be readable and debuggable without compile. Compile is an opt-in acceleration layer, never the default.
5. **Configuration, not code.** Users enable compile via `ServerArgs` flags (`--compile-level none|partial|full`), not by modifying model source.

### Phase 1: Partial compile (auxiliary modules)

**Goal**: Each model registers compile-eligible auxiliary modules. The framework compiles only those, leaving the backbone (transformer + RadixAttention) in eager/CUDA-graph mode.

**Framework API**:

Models expose compile targets via a protocol method:

```python
# Protocol that models can implement
class CompilableModel(Protocol):
    def get_compile_targets(self) -> dict[str, Callable]:
        """Return named compile targets: {name: callable}.

        Each callable must be fullgraph-compatible (tensor-only args,
        no dynamic control flow, no graph breaks).
        Returns empty dict if no auxiliary modules are compilable.
        """
        ...
```

Example implementation in S2-Pro:

```python
# sglang_model.py
class S2ProSGLangTextModel(nn.Module):
    def get_compile_targets(self) -> dict[str, Callable]:
        if not self._vq_ready:
            return {}
        return {"decode_codebooks": self._decode_codebooks_impl}
```

Where `_decode_codebooks_impl` is a standalone function extracted from the current `_decode_codebooks` method — pure tensor interface, no `self._output_codes[:bs]` slicing:

```python
# Top-level function, outside the class
def _decode_codebooks_impl(
    logits: Tensor,
    hidden_states: Tensor,
    semantic_bias: Tensor,
    audio_decoder: nn.Module,
    codebook_size: int,
    num_codebooks: int,
    semantic_begin_id: int,
) -> tuple[Tensor, Tensor]:
    """Returns (output_codes, semantic_ids). Compilable with fullgraph=True."""
    ...
    return output_codes, semantic_ids
```

**Framework-side wiring** (in engine factory layer, not model code):

```python
# engines/omni/compile.py  (new file, framework-level)
def apply_compile_targets(
    model: nn.Module,
    compile_mode: str = "max-autotune-no-cudagraphs",
) -> list[str]:
    """Compile all registered targets. Returns list of compiled target names."""
    if not hasattr(model, "get_compile_targets"):
        return []
    compiled = []
    for name, fn in model.get_compile_targets().items():
        compiled_fn = torch.compile(fn, mode=compile_mode, fullgraph=True)
        setattr(model, f"_compiled_{name}", compiled_fn)
        compiled.append(name)
    return compiled
```

The model's `forward()` calls the function pointer, which is either eager or compiled depending on whether `apply_compile_targets` was called:

```python
# In forward(), model calls whichever function pointer is set
codes, sem_ids = self._decode_codebooks_fn(...)
self._output_codes[:bs] = codes
self._output_semantic_ids[:bs] = sem_ids
```

**Compile mode**: `max-autotune-no-cudagraphs` — the framework owns CUDA graph capture; `torch.compile` handles kernel fusion only.

**What each model gains**:
- **S2-Pro**: Codebook loop compiled → ~121 tok/s (+37%)
- **Qwen3-Omni**: Talker decode, audio encoder forward → potential gains TBD (need benchmarks)
- **Future models**: Register their auxiliary modules, get compile for free

**Expected results (S2-Pro)**: ~121 tok/s, ~54s startup.

---

### Phase 2: Global compile (full model forward)

**Goal**: Compile the entire `model.forward()` for the remaining throughput tail, for use cases where startup time is acceptable (or solved by Phase 3).

**Why after Phase 1**: Global compile requires the transformer + RadixAttention path to be compile-clean. This is a larger surface area with potential graph breaks in:
- `RadixAttention` (paged KV cache management, dynamic token pools)
- `ForwardBatch` (Python dataclass with dynamic attributes)
- `MergedColumnParallelLinear` (tensor parallelism logic)
- RoPE cache access patterns

Phase 1 delivers ~85% of the compile gain immediately while Phase 2 works through these framework-level compatibility issues.

**Configuration surface** (exposed via `ServerArgs` for all models):

```python
# New ServerArgs field
compile_level: Literal["none", "partial", "full"] = "none"
```

- `none`: CUDA graph only (current default, zero startup cost)
- `partial`: Phase 1 — compile registered auxiliary modules only
- `full`: Phase 2 — `torch.compile` the entire `model.forward()`

**CUDA graph + torch.compile coexistence** — two strategies to benchmark:

1. **Layered**: `torch.compile(mode="max-autotune-no-cudagraphs")` for kernel fusion + our CUDA graph capture on top. Framework continues to own the graph.
2. **Unified**: `torch.compile(mode="reduce-overhead")` which manages its own CUDA graphs. Simpler if it works, but we lose fine-grained control over graph capture timing (e.g., S2-Pro's deferred capture after `setup_vq_decode`).

Both need benchmarking. Layered is the safer bet; unified is the cleaner end state if compatible.

**Expected results (S2-Pro)**: ~126 tok/s, ~137s startup (to be solved by Phase 3).

---

### Phase 3: Mega cache (framework-level compile artifact caching)

**Goal**: Eliminate startup overhead from Phase 1 and Phase 2. Second launch should be as fast as no-compile.

**Why last**: Cache infrastructure is a framework-level concern that benefits every model equally. It makes sense to build once the compile targets and levels are stable across models.

**Cache key design**:

```python
cache_key = hash(
    model_path,          # weights identity
    compile_level,       # "partial" or "full"
    max_batch_size,      # affects CUDA graph shapes
    torch.__version__,   # inductor codegen changes across versions
    cuda_runtime_version,
    gpu_arch,            # sm_80 vs sm_90 produce different kernels
    # per-model: any config that changes the compiled graph
    # (e.g., num_codebooks, codebook_size for S2-Pro)
)
```

Cache root: configurable via `--compile-cache-dir`, default `~/.cache/sglang-omni/compile_cache/`.

**Implementation options**:

1. **Inductor-native**: Set `TORCHINDUCTOR_CACHE_DIR` to our managed directory. Inductor already caches FX graphs and compiled kernels. We just need to manage the directory lifecycle and invalidation.
2. **`torch.export` + `torch._inductor.aot_compile`**: Pre-compile to `.so` artifacts at build/warmup time, load at startup. More control, but more complex.

Option 1 is simpler and should be tried first. Option 2 is the fallback if inductor's built-in cache doesn't provide sufficient speedup.

**Invalidation**: Recompile when any cache key component changes. CLI flag `--clear-compile-cache` for manual invalidation.

**Expected startup with warm cache**: Near baseline (~33s) even with `compile_level=full`.

## User-facing API (end state)

```bash
# Default: CUDA graph only, zero compile overhead
python -m sglang_omni.launch --model fishaudio/s2-pro

# Partial compile: fast head only, +20s startup, +37% throughput
python -m sglang_omni.launch --model fishaudio/s2-pro --compile-level partial

# Full compile: everything, +100s first startup, +43% throughput
python -m sglang_omni.launch --model fishaudio/s2-pro --compile-level full

# Full compile + warm cache: near-instant startup after first run
python -m sglang_omni.launch --model fishaudio/s2-pro --compile-level full \
    --compile-cache-dir /persistent/cache

# Works for any model
python -m sglang_omni.launch --model qwen3-omni --compile-level partial
```

## Non-goals (for now)

- **Model-specific compile hacks**: No `# torch.compile workaround` in model code. Fix the root cause.
- **Per-layer compile granularity**: Overkill. If needed, it's a Phase 2 refinement.
- **Compile for prefill path**: Prefill uses chunked processing, is not the throughput bottleneck, and has more dynamic shapes. Decode-only compile is sufficient.
- **Dynamic shape support**: All compile targets assume static shapes within a CUDA graph capture. `torch.compile` with dynamic shapes is a separate, harder problem.

## Dependencies

- [x] Unified slow+fast head refactor for S2-Pro (current `s2-optimize` branch)
- [ ] **Phase 1**: `audio_decoder.forward_kvcached` must be `fullgraph=True`-compatible
- [ ] **Phase 2**: SGLang's `RadixAttention`, `ForwardBatch`, `MergedColumnParallelLinear` must be compile-clean (may require upstream SGLang patches)
- [ ] **Phase 3**: Framework-level cache directory management (shared across all models)

## References

- [PR #153](https://github.com/sgl-project/sglang-omni/pull/153) — torch.compile benchmarking discussion with [@Ratish1](https://github.com/Ratish1)
- [PyTorch Inductor cache docs](https://pytorch.org/docs/stable/torch.compiler_inductor_profiling.html)
- [torch.compile + CUDA graphs interaction](https://pytorch.org/docs/stable/torch.compiler_cudagraph_trees.html)
- #172 — original issue (this issue generalizes the design to framework-level)

**Comment by yxs** [comment]
Hi @Ratish1, I am trying to reproduce your baseline numbers.

I checked out your exact commit (20be04a), ran the same benchmark script with `--no-ref-audio --max-samples 50`, and got **79.2 tok/s** baseline — vs your reported **88.1 tok/s**.



Could you share your environment details? Want to understand the ~10% gap.

```bash
jobuser [ ~/sglang-omni ]$   nvidia-smi | head -3
  python -c "import torch; print('torch:', torch.__version__, 'cuda:', torch.version.cuda)"
  python -c "import flash_attn; print('flash_attn:', flash_attn.__version__)" 2>/dev/null
  python --version
  cat /etc/os-release | head -2
Sat Mar 28 20:33:19 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.133.20             Driver Version: 570.133.20     CUDA Version: 13.0     |
torch: 2.9.1.7+cu130 cuda: 13.0
flash_attn: 2.8.3.1
Python 3.12.6
NAME="Microsoft Azure Linux"
VERSION="3.0.20250822"
```

```bash
#server
python -m sglang_omni.cli.cli serve --model-path /home/jobuser/models/s2-pro --port 8000 --config examples/configs/s2pro_tts.yaml

#benchmark
python -m benchmarks.benchmark_tts_speed --model fishaudio/s2-pro --port 8000 --testset /home/jobuser/data/seedtts_testset/en/meta.lst --max-samples 50 --no-ref-audio --output-dir results/baseline_ratish1_commit
```

```
============================================================
                    TTS Benchmark Result
============================================================
  Model:                         fishaudio/s2-pro
  Completed requests:            50
  Failed requests:               0
------------------------------------------------------------
  Latency mean (s):              1.374
  Latency median (s):            1.262
  Latency p95 (s):               2.32
  Latency p99 (s):               3.389
  RTF mean:                      0.2958
  RTF median:                    0.2941
  Audio duration mean (s):       4.69
  Tok/s (per-req mean):          79.0
  Tok/s (per-req median):        79.1
  Tok/s (aggregate):             79.2
  Gen tokens (mean):             101
  Gen tokens (total):            5049
  Throughput (req/s):            0.728
============================================================
```

**Comment by Hayden727** [comment]
> Hi @zhaochenyang20 — I'd like to take this.

**Tentative scope:** Phase 1 + Phase 3 in one PR, Phase 2 deferred.

- **Phase 1 — partial compile.**  
  Implement `CompilableModel.get_compile_targets() -> dict[str, Callable]` and `apply_compile_targets()` in a new `engines/omni/compile.py` exactly as sketched in the issue:  
  `mode="max-autotune-no-cudagraphs"`, `fullgraph=True`, framework-owned CUDA-graph capture stays untouched.  
  First adopter is S2-Pro's `_decode_codebooks_impl` (factored out as a pure tensor-in/tensor-out function); Qwen3-Omni's code predictor and code2wav follow once the protocol is in.

- **Phase 3 — mega cache.**  
  Start with Option 1 (`TORCHINDUCTOR_CACHE_DIR` under a managed root), keyed exactly as proposed: `(model_path, compile_level, max_batch_size, torch.__version__, cuda_runtime, gpu_arch)`.  
  Add `--compile-cache-dir` and `--clear-compile-cache`.  
  Fall back to `torch.compiler.save_cache_artifacts` only if Option 1 can't survive process restarts cleanly.

- **Phase 2 — deferred.**  
  The benchmark row in the issue (~121 → ~126 tok/s for ~2.5× startup, plus the RadixAttention / ForwardBatch / TP compile-clean work) argues for landing the framework surface + cache first and revisiting once Phase 3 hides the remaining cost. I can fold it in if you'd rather have everything at once — just let me know.

**Respecting the design principles:**  
no `torch.compile()` inside model files, targets are self-contained pure-tensor callables, and `--compile-level=none` remains the default so the eager path stays first-class.
